# Emoji Sentiment

Are popular emojis generally associated with positive or negative sentiments?

The file `"emoji-sentiment.csv"` provides data on the sentiment associated with various emojis.

Researchers examined 1.6 million tweets across 13 European languages. Each tweet was labeled by annotators as positive (+1), negative (-1), or neutral (0). About 4% of these tweets included emojis.

Columns include:
- `Occurrences [5...max]`: Number of times the emoji appears in the dataset.
- `Position [0...1]`: Average position of the emoji in tweets, from start (0) to end (1).
- `Neg [0...1]`: Percentage of tweets with the emoji that are 'negative'.
- `Neu [0...1]`: Percentage of tweets with the emoji that are 'neutral'.
- `Pos [0...1]`: Percentage of tweets with the emoji that are 'positive'.



In [1]:
# FOR GOOGLE COLAB ONLY.
# Uncomment and run the code below. A dialog will appear to upload files.
# Upload 'emoji-sentiment.csv'.

# from google.colab import files
# uploaded = files.upload()

In [2]:
import pandas as pd
df = pd.read_csv('emoji-sentiment.csv')
df.head(3)

,Char,Image [twemoji],Unicode codepoint,Occurrences [5...max],Position [0...1],Neg [0...1],Neut [0...1],Pos [0...1],Sentiment bar (c.i. 95%),Unicode name,Unicode block
0,😂,😂,0x1f602,14622,0.805,0.247,0.285,0.468,NaN,FACE WITH TEARS OF JOY,Emoticons
1,❤,❤,0x2764,8050,0.747,0.044,0.166,0.790,NaN,HEAVY BLACK HEART,Dingbats
2,♥,♥,0x2665,7144,0.754,0.035,0.272,0.693,NaN,BLACK HEART SUIT,Miscellaneous Symbols


### Project Ideas:

Data Cleaning: 
- Remove unnecessary columns that are not useful for your analysis.

- Rename the remaining columns using `snake_case` (all lowercase letters with underscores between words).

New Variables:
- Add a new column called `sentiment`, where sentiment = (% positive tweets) - (% negative tweets).

- Add a `positive_flag` column that is `True` if `sentiment > 0` (or above a set threshold), otherwise `False`.

Types of questions you can now answer more easily:
- What percentage of emojis in the dataset have a positive sentiment?

- What percentage of the top 20 most popular emojis are positive?

- Which emoji (with more than 500 mentions) is the most positive?

- Which emoji (with more than 500 mentions) is the most negative?

- Where in the tweets are most emojis located (i.e. at the beginning or the end)?

- Is there a difference in the placement of positive versus negative emojis within a tweet?

**DC1. Remove unnecessary columns that are not useful for your analysis.**

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 751 entries, 0 to 750
Data columns (total 11 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Char                      751 non-null    str    
 1   Image [twemoji]           751 non-null    str    
 2   Unicode codepoint         751 non-null    str    
 3   Occurrences [5...max]     751 non-null    int64  
 4   Position [0...1]          751 non-null    float64
 5   Neg [0...1]               751 non-null    float64
 6   Neut [0...1]              751 non-null    float64
 7   Pos [0...1]               751 non-null    float64
 8   Sentiment bar (c.i. 95%)  0 non-null      float64
 9   Unicode name              751 non-null    str    
 10  Unicode block             751 non-null    str    
dtypes: float64(5), int64(1), str(5)
memory usage: 64.7 KB


Sentiment Bar consists 0 non-null (all are NaN), we can remove this column. Also, unicode codepoint won't be needed for the analysis it can be removed as well.

In [4]:
emoji = df.drop(columns = ['Sentiment bar (c.i. 95%)', 'Unicode codepoint'])
emoji

,Char,Image [twemoji],Occurrences [5...max],Position [0...1],Neg [0...1],Neut [0...1],Pos [0...1],Unicode name,Unicode block
0,😂,😂,14622,0.805,0.247,0.285,0.468,FACE WITH TEARS OF JOY,Emoticons
1,❤,❤,8050,0.747,0.044,0.166,0.790,HEAVY BLACK HEART,Dingbats
2,♥,♥,7144,0.754,0.035,0.272,0.693,BLACK HEART SUIT,Miscellaneous Symbols
3,😍,😍,6359,0.765,0.052,0.219,0.729,SMILING FACE WITH HEART-SHAPED EYES,Emoticons
4,😭,😭,5526,0.803,0.436,0.220,0.343,LOUDLY CRYING FACE,Emoticons
...,...,...,...,...,...,...,...,...,...
746,♮,♮,5,0.937,0.125,0.625,0.250,MUSIC NATURAL SIGN,Miscellaneous Symbols
747,🅾,🅾,5,0.977,0.375,0.375,0.250,NEGATIVE SQUARED LATIN CAPITAL LETTER O,Enclosed Alphanumeric Supplement
748,🔄,🔄,5,0.971,0.125,0.750,0.125,ANTICLOCKWISE DOWNWARDS AND UPWARDS OPEN CIRCL...,Miscellaneous Symbols and Pictographs
749,☄,☄,5,0.435,0.125,0.750,0.125,COMET,Miscellaneous Symbols


**DC2: Rename the remaining columns using `snake_case` (all lowercase letters with underscores between words).**

In [5]:
name_map = {
    'Char' : 'char',
    'Image [twemoji]' : 'image',
    'Occurrences [5...max]' : 'occurences',
    'Position [0...1]' : 'position',
    'Neg [0...1]' : 'negative',
    'Neut [0...1]' : 'neutral',
    'Pos [0...1]' : 'positive',
    'Unicode name' : 'unicode_name',
    'Unicode block' : 'unicode_block'
}

emoji = emoji.rename(columns = name_map)
emoji

,char,image,occurences,position,negative,neutral,positive,unicode_name,unicode_block
0,😂,😂,14622,0.805,0.247,0.285,0.468,FACE WITH TEARS OF JOY,Emoticons
1,❤,❤,8050,0.747,0.044,0.166,0.790,HEAVY BLACK HEART,Dingbats
2,♥,♥,7144,0.754,0.035,0.272,0.693,BLACK HEART SUIT,Miscellaneous Symbols
3,😍,😍,6359,0.765,0.052,0.219,0.729,SMILING FACE WITH HEART-SHAPED EYES,Emoticons
4,😭,😭,5526,0.803,0.436,0.220,0.343,LOUDLY CRYING FACE,Emoticons
...,...,...,...,...,...,...,...,...,...
746,♮,♮,5,0.937,0.125,0.625,0.250,MUSIC NATURAL SIGN,Miscellaneous Symbols
747,🅾,🅾,5,0.977,0.375,0.375,0.250,NEGATIVE SQUARED LATIN CAPITAL LETTER O,Enclosed Alphanumeric Supplement
748,🔄,🔄,5,0.971,0.125,0.750,0.125,ANTICLOCKWISE DOWNWARDS AND UPWARDS OPEN CIRCL...,Miscellaneous Symbols and Pictographs
749,☄,☄,5,0.435,0.125,0.750,0.125,COMET,Miscellaneous Symbols


**NV1: Add a new column called `sentiment`, where sentiment = (% positive tweets) - (% negative tweets).**

In [9]:
emoji['sentiment'] = emoji.eval('positive - negative')
emoji

,char,image,occurences,position,negative,neutral,positive,unicode_name,unicode_block,sentiment
0,😂,😂,14622,0.805,0.247,0.285,0.468,FACE WITH TEARS OF JOY,Emoticons,0.221
1,❤,❤,8050,0.747,0.044,0.166,0.790,HEAVY BLACK HEART,Dingbats,0.746
2,♥,♥,7144,0.754,0.035,0.272,0.693,BLACK HEART SUIT,Miscellaneous Symbols,0.658
3,😍,😍,6359,0.765,0.052,0.219,0.729,SMILING FACE WITH HEART-SHAPED EYES,Emoticons,0.677
4,😭,😭,5526,0.803,0.436,0.220,0.343,LOUDLY CRYING FACE,Emoticons,-0.093
...,...,...,...,...,...,...,...,...,...,...
746,♮,♮,5,0.937,0.125,0.625,0.250,MUSIC NATURAL SIGN,Miscellaneous Symbols,0.125
747,🅾,🅾,5,0.977,0.375,0.375,0.250,NEGATIVE SQUARED LATIN CAPITAL LETTER O,Enclosed Alphanumeric Supplement,-0.125
748,🔄,🔄,5,0.971,0.125,0.750,0.125,ANTICLOCKWISE DOWNWARDS AND UPWARDS OPEN CIRCL...,Miscellaneous Symbols and Pictographs,0.000
749,☄,☄,5,0.435,0.125,0.750,0.125,COMET,Miscellaneous Symbols,0.000


**NV2: Add a `positive_flag` column that is `True` if `sentiment > 0` (or above a set threshold), otherwise `False`.**

In [11]:
emoji['positive_flag'] = emoji.eval('sentiment > 0')
emoji

,char,image,occurences,position,negative,neutral,positive,unicode_name,unicode_block,sentiment,positive_flag
0,😂,😂,14622,0.805,0.247,0.285,0.468,FACE WITH TEARS OF JOY,Emoticons,0.221,True
1,❤,❤,8050,0.747,0.044,0.166,0.790,HEAVY BLACK HEART,Dingbats,0.746,True
2,♥,♥,7144,0.754,0.035,0.272,0.693,BLACK HEART SUIT,Miscellaneous Symbols,0.658,True
3,😍,😍,6359,0.765,0.052,0.219,0.729,SMILING FACE WITH HEART-SHAPED EYES,Emoticons,0.677,True
4,😭,😭,5526,0.803,0.436,0.220,0.343,LOUDLY CRYING FACE,Emoticons,-0.093,False
...,...,...,...,...,...,...,...,...,...,...,...
746,♮,♮,5,0.937,0.125,0.625,0.250,MUSIC NATURAL SIGN,Miscellaneous Symbols,0.125,True
747,🅾,🅾,5,0.977,0.375,0.375,0.250,NEGATIVE SQUARED LATIN CAPITAL LETTER O,Enclosed Alphanumeric Supplement,-0.125,False
748,🔄,🔄,5,0.971,0.125,0.750,0.125,ANTICLOCKWISE DOWNWARDS AND UPWARDS OPEN CIRCL...,Miscellaneous Symbols and Pictographs,0.000,False
749,☄,☄,5,0.435,0.125,0.750,0.125,COMET,Miscellaneous Symbols,0.000,False


**QUESTION 1: What percentage of emojis in the dataset have a positive sentiment?**

In [61]:
emoji['positive_flag'].mean() * 100

np.float64(82.42343541944075)

**ANSWER** = 82.42%. (This code only count the "True" values. Since True = 1 and False = 0.)

**QUESTION 2: What percentage of the top 20 most popular emojis are positive?**

In [48]:
top_20 = emoji.sort_values('occurences', ascending=False).head(20)
top_20['positive_flag'].sum() / len(top_20) * 100

np.float64(90.0)

**ANSWER** = 90% of the top 20 most popular emojis on tweets are flagged positive.

**QUESTION 3: Which emoji (with more than 500 mentions) is the most positive?**

In [59]:
emo_500 = emoji.query('occurences > 500')
emo_500.sort_values('sentiment', ascending=False).head(1)

,char,image,occurences,position,negative,neutral,positive,unicode_name,unicode_block,sentiment,positive_flag
1,❤,❤,8050,0.747,0.044,0.166,0.79,HEAVY BLACK HEART,Dingbats,0.746,True


**ANSWER** = HEAVY BLACK HEART

**QUESTION 4: Which emoji (with more than 500 mentions) is the most negative?**

In [60]:
emo_500 = emoji.query('occurences > 500')
emo_500.sort_values('sentiment', ascending=True).head(1)

,char,image,occurences,position,negative,neutral,positive,unicode_name,unicode_block,sentiment,positive_flag
23,😒,😒,1385,0.858,0.591,0.192,0.217,UNAMUSED FACE,Emoticons,-0.374,False


**ANSWER** = UNAMUSED FACE

**QUESTION 5: Where in the tweets are most emojis located (i.e. at the beginning or the end)?**

In [63]:
emoji['position'].mean()

np.float64(0.6655486018641811)

**ANSWER** = Since a value near to 0 means that the position is at the beginning of a tweet(and 1 at the end). The resulting value 0.66555 (around 0.5) indicates that most emojis are located at the middle of the tweets.

**QUESTION 6: Is there a difference in the placement of positive versus negative emojis within a tweet?**

In [79]:
#positive average
emoji.query('positive_flag == True')['position'].mean()

np.float64(0.662248788368336)

In [80]:
#negative average
emoji.query('positive_flag == False')['position'].mean()

np.float64(0.6810227272727273)

**ANSWER** = There is an approximately 0.02 difference in the position of positive versus negative emojis within a tweet. The values also imply that regardless of the sentiment value of the emojis, they are mostly placed at the middle of a tweet.